Update HTTP method testing. (#853)
OWASP/wstg: document/4-Web_Application_Security_Testing/02-Configuration_and_Deployment_Management_Testing/06-Test_HTTP_Methods.md

## Summary

HTTP offers a number of methods (or verbs) that can be used to perform actions on the web server. While GET and POST are by far the most common methods that are used to access information provided by a web server, there are a variety of other methods that may also be supported, and can sometimes be exploited by attackers.

[RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231) defines the main valid HTTP request methods (or verbs), although additional methods have been added in other RFCs, such as [RFC 5789](https://datatracker.ietf.org/doc/html/rfc5789). Several of these verbs have bee re-used for different purposes in [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) applications, listed in the table below.

| Method | Original Purpose | RESTful Purpose |
|--------|------------------|-----------------|
| [`GET`](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.1) | Request a file. | Request an object.|
| [`HEAD`](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.2) | Request a file, but only return the HTTP headers. | |

> | [`POST`](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.3) | Submit data. | |

Reviewer: How about we set here for POST "Add to an object" in restful?
It happens mostly in 2 cases:
1. Add items to for example a json object
2. Add a new entity to a parent `POST /api/people -- { <data here for new object> }`

Author: I'm not sure that's really accurate though? A REST API would normally use a `POST` request for a login page (for instance) - but that's not adding an item to an object, or creating a new entity.

Reviewer: Hmmm, yeah. Let's stay without it since it serves so many purposes.

> | [`PUT`](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.4) | Upload a file. | Create an object. 

Reviewer 2: ```suggestion
| [`PUT`](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.4) | Upload a file. | Create an object. |
```

I believe the lack of trailing pipe is the issue the linter is having.

### Discover the Supported Methods

To perform this test, the tester needs some way to identify which HTTP methods are supported by the web server that is being examined. The simplest way to do this is to make an `OPTIONS` request to the server:

```http
OPTIONS / HTTP/1.1
Host: example.org
```

The server should then response with a list of supported methods:

```http
HTTP/1.1 200 OK
Allow: OPTIONS, GET, HEAD, POST
```

However, some servers may not respond to `OPTIONS` requests, or may return inaccurate information. Additionally, servers may support different methods for different paths - so just because a method is not supported for the root `/` directory, this doesn't necessarily mean that it won't be supported elsewhere.

> An alternative way to test for supported methods is to simply make a request with that method type, and examine the server response. If the method is not permitted, the server should return a `405 Method Not Allowed` status.

Reviewer: I feel like removing "alternative way". Sometimes OPTIONS requests aren't truthful, or default set by a server, etc. So maybe a word of advice that test things out always?

Author: Changed to "a more reliable way"

Note that some servers treat unknown methods as equivalent to `GET`, so they may respond to arbitrary methods, such as the request shown below. This can occasionally be useful to evade a web application firewall, or any other filtering that blocks specific methods.

```http
FOO / HTTP/1.1
Host: example.org
```

Requests with arbitrary methods can also be made using cURL with the `-X` option:

```bash
curl -X FOO https://example.org
```

> There are also a variety of automated tools that can attempt to determine supported methods, such as the [`http-methods`](https://nmap.org/nsedoc/scripts/http-methods.html) Nmap script. However, these tools may not test for dangerous methods (such as `PUT` or `DELETE`), or may unintentionally cause changes to the web server if these methods are supported. As such, they should be used with care.

Reviewer: Hmm, we know about "dangerous methods", but new readers won't. Should we add a link to safe methods?
https://developer.mozilla.org/en-US/docs/Glossary/Safe/HTTP is an example

Author: Changed this to:

...dangerous methods (i.e, methods that may cause changes such as `PUT`...)

Reviewer: Mind if we add that to the references at the end? The updated sentence explains it well, just want to be sure that users have a clear reference for it.

Author: Added

The `PUT` and `DELETE` methods can have different effects, depending on whether they are being interpreted by the web server or by the application running on it.

#### Legacy Web Servers

Some legacy web servers allowed the use of the `PUT` method to create files on the server. For example, if the server is configured to allow this, the request below would create a file on the server called `test.html` with the contents `<script>alert(1)</script>`.

```http
PUT /test.html HTTP/1.1
Host: example.org
Content-Length: 25

<script>alert(1)</script>
```

Similar requests can also be made with cURL:

```bash
curl https://example.org --upload-file test.html
```

This allows an attacker to upload arbitrary files to the webserver, which could potentially result in a full system compromise if they are allowed to upload executable code such as PHP files. However, this configuration is extremely rare, and is unlikely to be seen on any modern systems.

Similarly, the `DELETE` method can be used to delete files from the webserver. Note that this is a **destructive action**, so care should be taken when testing this method.

```http
DELETE /test.html HTTP/1.1
Host: example.org
```

> Or with cURL

Reviewer: Really picky comment, but let's add a `:` at the end

```bash
curl http://example.org/test.html -X DELETE
```

#### RESTful APIs

By contrast, the `PUT` and `DELETE` methods are commonly used by modern RESTful applications to create and delete objects. For example, the API request below could be used to create a user called "foo" with a role of "user":

```http
PUT /api/users/foo HTTP/1.1
Host: example.org
Content-Length: 34

{
    "role": "user"
}
```

A similar request with the DELETE method could be used to delete an object.

```http
DELETE /api/users/foo HTTP/1.1
Host: example.org
```

> Although it may be reported by automated scanning tools, the presence of these methods on a RESTful API **is not a security issue**. However, this functionality may have other vulnerabilities (such as weak access control), should be thoroughly tested.

Reviewer: ```suggestion
Although it may be reported by automated scanning tools, the presence of these methods on a RESTful API **is not a security issue**. However, this functionality may have other vulnerabilities (such as weak access control), and should be thoroughly tested.
```

For example, if we have an user called "foo" with the following properties:

```json
{
    "role": "user",
    "email": "foo@example.org"
}
```

The following JSON PATCH request could be used to change the role of this user "admin", without modifying the email address:

```http
PATCH /api/users/foo HTTP/1.1
Host: example.org

{ "op": "replace", "path": "/role", "value": "admin" }
```

Although the RFC states that it should include instructions for how the object should be modified, the `PATCH` method is commonly (mis)used to include the changed content instead, as shown below. Much like the previous request, this would change the "role" value to "admin" without modifying the rest of the object. This is in contrast to the `PUT` method, which would overwrite the entire object (and thus result in an object with no "email" attribute).

```http
PATCH /api/users/foo HTTP/1.1
Host: example.org

{
    "role": "admin"
}
```

> As with the `PUT` method, this functionality may have access control weaknesses or other vulnerabilities. Additionally, applications may not performed the same level of input validation when modifying an object as they do when creating one. This could potentially allow malicious values to be injected (such as in a stored cross-site scripting attack), or could allow broken or invalid objects that may result in business logic related issues.

Reviewer: ```suggestion
As with the `PUT` method, this functionality may have access control weaknesses or other vulnerabilities. Additionally, applications may not perform the same level of input validation when modifying an object as they do when creating one. This could potentially allow malicious values to be injected (such as in a stored cross-site scripting attack), or could allow broken or invalid objects that may result in business logic related issues.
```

### Testing for Access Control Bypass

> If a page on the application redirects users to a login page with a `302` code when they try and access it direct, it may be possible to bypass this by making a request with a different HTTP method, such as `HEAD`, `POST` or even a made up method such as `FOO`. If the web application responds with a `HTTP/1.1 200 OK` rather than the expected `HTTP/1.1 302 Found` then it may be possible to bypass the authentication or authorization. The example below shows how a `HEAD` request may result in a page setting administrative cookies, rather than redirecting the user to a login page:

Reviewer: Other than the suggestion, this feels a bit repetitive. Not a hard requirement to merge, just an observation.
```suggestion
If a page on the application redirects users to a login page with a `302` code when they attempt to access it directly, it may be possible to bypass this by making a request with a different HTTP method, such as `HEAD`, `POST` or even a made up method such as `FOO`. If the web application responds with a `HTTP/1.1 200 OK` rather than the expected `HTTP/1.1 302 Found` then it may be possible to bypass the authentication or authorization. The example below shows how a `HEAD` request may result in a page setting administrative cookies, rather than redirecting the user to a login page:
```

```http
HEAD /admin/ HTTP/1.1
Host: example.org
```

```http
HTTP/1.1 200 OK
[...]
Set-Cookie: adminSessionCookie=[...];
```

> Alternatively, it may be possible make direct requests to pages that cause actions, such as:

Reviewer 2: ```suggestion
Alternatively, it may be possible to make direct requests to pages that cause actions, such as:
```

### Discover the Supported Methods

To perform this test, the tester needs some way to identify which HTTP methods are supported by the web server that is being examined. The simplest way to do this is to make an `OPTIONS` request to the server:

```http
OPTIONS / HTTP/1.1
Host: example.org
```

> The server should then response with a list of supported methods:

Reviewer 2: Need a blank line in here to address the linter issue.

### Discover the Supported Methods

To perform this test, the tester needs some way to identify which HTTP methods are supported by the web server that is being examined. The simplest way to do this is to make an `OPTIONS` request to the server:

```http
OPTIONS / HTTP/1.1
Host: example.org
```
The server should then response with a list of supported methods:

```http
HTTP/1.1 200 OK
Allow: OPTIONS, GET, HEAD, POST
```

However, some servers may not respond to `OPTIONS` requests, or may return inaccurate information. Additionally, servers may support different methods for different paths - so just because a method is not supported for the root `/` directory, this doesn't necessarily mean that it won't be supported elsewhere.

> An alternatively way to test for supported methods is to simply make a request with that method type, and examine the server response. If the method is not permitted, the server should return a `405 Method Not Allowed` status.

Reviewer 2: alternatively ➡️ alternative

> Note that some servers treat unknown methods as equivalent to `GET`, so they may response to arbitrary methods, such as the request shown below. This can occasionally be useful to evade a web application firewall, or any other filtering that blocks specific methods.

Reviewer 2: may response ➡️ may respond

```http
FOO / HTTP/1.1
Host: example.org
```

Requests with arbitrary methods can also be made using cURL with the `-X` option:

```bash
curl -X FOO https://example.org
```

> There are also a variety of automated tools that can attempt to determine supported methods, such as the [`http-methods`](https://nmap.org/nsedoc/scripts/http-methods.html) Nmap script. However, these tools may not test for dangerous methods (such as `PUT` or `DELETE`), or may unintentionally cause changes to the web server if this methods are supported. As such, they should be used with care.

Reviewer 2: if this methods ➡️ if these methods

### PUT and DELETE

The `PUT` and `DELETE` methods can have different effects, depending on whether they are being interpreted by the web server or by the application running on it.

#### Legacy Web Servers

Some legacy web servers allowed the use of the `PUT` method to create files on the server. For example, if the server is configured to allow this, the request below would create a file on the server called `test.html` with the contents `<script>alert(1)</script>`.

```http
PUT /test.html HTTP/1.1
Host: example.org
Content-Length: 25

<script>alert(1)</script>
```

Similar requests can also be made with cURL:

```bash
curl https://example.org --upload-file test.html
```

> This allows an attacker to upload arbitrary files to the webserver, which could potential result in a full system compromise if they are allowed to upload executable code such as PHP files. However, this configuration is extremely rare, and is unlikely to be seen on any modern systems.

Reviewer 2: could potential ➡️ could potentially

Reviewer 2: > code such as PHP files.

Worth mentioning something about web shells, rce, whatever? (Just as slightly more concrete context.)

Author: I feel like the existing text is fairly explicit? It says that you can upload PHP code to compromise a system - surely anyone who's doing this kind of testing should understand that?

Reviewer 2: I'm fine with that. We've generally been assuming close to zero knowledge, but I also overthink things sometimes. So let's leave it as-is. We can always tweak later if someone wonders/asks. You're right it shouldn't be much of a leap if you're already reading this stuff :wink:

Similarly, the `DELETE` method can be used to delete files from the webserver. Note that this is a **destructive action**, so care should be taken when testing this method.

```http
DELETE /test.html HTTP/1.1
Host: example.org
```

Or with cURL

```bash
curl http://example.org/test.html -X DELETE
```

#### RESTful APIs

> By contrast, the `PUT` and `DELETE` methods are commonly used by modern RESTful applications to create and delete objects. For example, the API request below could be use to create a user called "foo" with a role of "user":

Reviewer 2: be use to create ➡️ be used to create

```http
PUT /api/users/foo HTTP/1.1
Host: example.org
Content-Length: 34

{
    "role": "user"
}
```

> A similar requests with the DELETE method could be used to delete an object.

Reviewer 2: requests ➡️ request

```http
DELETE /api/users/foo HTTP/1.1
Host: example.org
```

Although it may be reported by automated scanning tools, the presence of these methods on a RESTful API **is not a security issue**. However, this functionality may have other vulnerabilities (such as weak access control), should be thoroughly tested.

### TRACE

> The `TRACE` method causes the server to echo back the contents of the request. This lead to a vulnerability called Cross-Site Tracing (XST) being published in [2003](https://www.cgisecurity.com/whitehat-mirror/WH-WhitePaper_XST_ebook.pdf) (PDF), which could be used to access cookies that had the `HttpOnly` flag set. The `TRACE` method has been blocked in all browsers and plugins for many years, and as such this issue is no longer exploitable. However, it may still be flagged by automated scanning tools, and the `TRACE` method being enabled on a web server suggests that is has not been properly hardened.

Reviewer 2: Is it worth noting that "TRACK" is MS's implementation?

### CONNECT

> The `CONNECT` method causes the web server to open a TCP connection to another system, and then to pass traffic from the client through to that system. This could allow an attacker to proxy traffic through the server, in order to hide their source address, access internal systems or access services that are bound to localhost. An example of a CONNECT request is shown below:

Reviewer 2: I dunno it isn't "wrong" in any way but to be more consistent with earlier content the second "CONNECT" should probably be code fenced as well.

```http
CONNECT 192.168.0.1:443 HTTP/1.1
Host: example.org
```

### PATCH

The `PATCH` method is defined in [RFC 5789](https://datatracker.ietf.org/doc/html/rfc5789), and is used to provide instructions for how an object should be modified. The RFC itself does not define what format these instructions should be in, but various methods are defined in other standards, such as the [RFC 6902 - JavaScript Object Notation (JSON) Patch](https://datatracker.ietf.org/doc/html/rfc6902).

For example, if we have an user called "foo" with the following properties:

```json
{
    "role": "user",
    "email": "foo@example.org"
}
```

> The following JSON PATCH request would change the role of this user "admin", without modifying the email address:

Reviewer 2: ```suggestion
The following JSON PATCH request could change the role of this user to "admin", without modifying the email address:
```

- [Ncat](https://nmap.org/ncat/)
- [cURL](https://curl.haxx.se/)

> - [Nmap http-methods NSE script](https://nmap.org/nsedoc/scripts/http-methods.html)

Reviewer 2: Thanks for tweaking this. I hadn't put any thought into it before, but this does seem to be correct from a quick check I just did. I'll look into implementing a textlint rule for this :wink:
